### PR TITLE
Api view comments

### PR DESCRIPTION
**Pull request by gearama**
#closes 5415
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/tables/azure-data-tables/inc/azure/data/tables.hpp [hunk 1]
```diff
@@ -12,5 +12,6 @@
 #include "azure/data/tables/dll_import_export.hpp"
 #include "azure/data/tables/models.hpp"
 #include "azure/data/tables/rtti.hpp"
+#include "azure/data/tables/table_sas_builder.hpp"
```
Comment (RickWinter) [inline sdk/tables/azure-data-tables/inc/azure/data/tables.hpp:15]: Tables plural for consistency?
```suggestion
#include "azure/data/tables/tables_sas_builder.hpp"

```

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp [hunk 2]
```diff
@@ -54,41 +53,45 @@ namespace Azure { namespace Data { namespace Tables {
...
      * the response body.
      */
-    enum class ListTablesIncludeFlags
+    enum class QueryTablesIncludeFlags
     {
       None = 0,
       Metadata = 1,
     };
-    inline ListTablesIncludeFlags operator|(ListTablesIncludeFlags lhs, ListTablesIncludeFlags rhs)
+    inline QueryTablesIncludeFlags operator|(
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:61]: Do you want to consider adding all the logical operators for completeness?

Also, if we want to enable logical operators on bitfield enumerations, this might be a situation to relax the rules and use a #define to define these operators (there's a Windows macro that does this which I've found helpful in the past).

If it's a one-off it shouldn't be a huge issue.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp [hunk 3]
```diff
@@ -54,41 +53,45 @@ namespace Azure { namespace Data { namespace Tables {
...
-      using type = std::underlying_type_t<ListTablesIncludeFlags>;
-      return static_cast<ListTablesIncludeFlags>(static_cast<type>(lhs) | static_cast<type>(rhs));
+      using type = std::underlying_type_t<QueryTablesIncludeFlags>;
+      return static_cast<QueryTablesIncludeFlags>(static_cast<type>(lhs) | static_cast<type>(rhs));
     }
-    inline ListTablesIncludeFlags& operator|=(
-        ListTablesIncludeFlags& lhs,
-        ListTablesIncludeFlags rhs)
+    inline QueryTablesIncludeFlags& operator|=(
+        QueryTablesIncludeFlags& lhs,
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:69]: Using a reference on POD (Plain Old Data) is often a mistake because it significantly hurts performance.

Does the `lhs` parameter really need to be passed by reference?

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp [hunk 4]
```diff
@@ -391,14 +409,34 @@ namespace Azure { namespace Data { namespace Tables {
...
 
     /**
-     * @brief List tables.
+     * @brief Create the table indicated in the tableName field of the client.
+     *
+     * @param context for canceling long running operations.
+     * @param tableName The name of the table to be created.
+     * @return Create table result.
+     */
+    Response<Models::Table> CreateTable(std::string tableName, Core::Context const& context = {});
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:418]: The `tableName` parameter should probably be declared as `const&` rather than being passed by value.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp [hunk 5]
```diff
@@ -391,14 +409,34 @@ namespace Azure { namespace Data { namespace Tables {
...
+
+    /**
+     * @brief Delete the table indicated in the tableName field of the client.
+     *
+     * @param context for canceling long running operations.
+     * @param tableName The name of the table to be deleted.
+     * @return Delete table result.
+     */
+    Response<Models::DeleteResult> DeleteTable(
+        std::string tableName,
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:428]: Same comment as above: In general, it's better to pass complex objects as `const&` rather than by value, because passing by value requires taking a copy of the object and destroying it afterwards.

The single exception to this is in a method constructor, where there is a school of thought to pass constructor structure parameters which will be used to initialize a struct member by value and then use std::move to move the temporary value to the struct member. 

I'm personally not a fan of this pattern because it pushes the code to construct the copy to the caller (thus every caller has the code to construct the copy) but that isn't relevant - the pattern exists and there are places which recommend it.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    /**
+     * @brief No protocol has been specified. If no value is specified,
+     * the service will default to HttpsAndHttp.
+     */
+    None = 0,
+
+    /**
+     * @brief Only requests issued over HTTPS or HTTP will be permitted.
+     */
+    HttpsAndHttp = 1,
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:31]: Is there actually a real-world scenario where we allow HTTP access to the Tables service?

I thought we had a hard and fast rule that no azure services allowed HTTP access.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+     * @brief The optional end of the partition key values range being made available.
+     */
+    std::string RowKeyEnd;
+
+    /**
+     * @brief Sets the permissions for the table SAS.
+     *
+     * @param permissions The allowed permissions.
+     */
+    void SetPermissions(TablesSasPermissions permissions);
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:154]: For better or worse, the convention we've used for `struct` is that a `struct` is used only when it has no methods (including constructor/explicit destructors). 

I'm personally not a fan of this pattern (I'd rather see `struct` be used in the C++ meaning of "a class which has only public methods" rather than the C meaning of "something that looks like plain old data, even if it's not"), but this IS the pattern we've settled on as a team.

Bottom line: This should be a class, not a struct :(.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+    /**
+     * @brief The optional end of the partition key values range being made available.
+     */
+    std::string PartitionKeyEnd;
+
+    /**
+     * @brief The optional start of the row key values range being made available.
+     */
+    std::string RowKeyStart;
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:142]: What is the syntax of the `RowKeyStart`, `PartitionKeyStart`, `PartitionKeyEnd`, etc? Are they free form text, or something else?

File: sdk/tables/azure-data-tables/samples/tables_getting_started.cpp [hunk 9]
```diff
@@ -33,17 +33,17 @@ int main()
   auto tableClient = TableClient::CreateFromConnectionString(GetConnectionString(), TableName);
 
   // create new table
-  tableClient.Create();
+  tableServiceClient.CreateTable(TableName);
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/samples/tables_getting_started.cpp:36]: Does this function return a value? Is it a best practice to check the return value, or is it going to throw an exception on failure?

File: sdk/tables/azure-data-tables/src/tables_clients.cpp [hunk 10]
```diff
@@ -135,6 +135,17 @@ TableServicesClient::TableServicesClient(
       std::move(perOperationPolicies2));
 }
 
+TableServicesClient::TableServicesClient(
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/src/tables_clients.cpp:138]: Is there more than one Table service? This class definition is `TableServicesClient`, with Services being plural. That implies that there is more than one service.

File: sdk/tables/azure-data-tables/test/ut/sas_test.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+#include <sstream>
+#include <string>
+#include <thread>
+
+using namespace Azure::Data::Tables::Sas;
+// cspell: words rwdxylacupitf raud bqft
+namespace Azure { namespace Data { namespace Test {
+  TEST(SasTest, TableSasBuilderTestAllSet)
+  {
+    TablesSasBuilder sasBuilder;
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: I'm going to ask an obnoxious question.

The `Builder` pattern is a fairly well known pattern where the builder class has a set of methods which (for instance) do things like:

```c++

TablesSasBuilder sasBuilder;
sasBuilder
  .SetPermissions(TablesSasPermissions::All)
  .SetStartsOn(...)
  .SetExpiresOn(...)
  .SetIdentifier("myIdentifier")
  .SetIPRange("iprange")

 :
 :
```

Typically classes which implement the "Builder" pattern have a suffix of "Builder".

When I saw `TablesSasBuilder`, I assumed that it implemented the Builder pattern, so I was a bit surprised to see that it didn't.

Is there a reason to call this a builder given that it doesn't follow the builder pattern? Possibly something like SasBuilderFactory (which would imply it followed the factory pattern which is a different pattern altogether).

Anyway, I'm not asking for a change here, but it IS something to think about. If (for instance), the other languages have a similar shape to this shape, I'm ok with this. But I'd like to understand how this type looks in all the existing languages.

**Review by LarryOsterman — COMMENTED** [review]
In general, it looks good, but I've left a bunch of questions that I'd like to see answered, and a couple of things to think about.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp (hunk 2 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:61]: what is this macro , i could use it,

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp (hunk 3 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:69]: done

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp [hunk 12]
```diff
@@ -347,6 +350,42 @@ namespace Azure { namespace Data { namespace Tables {
       std::string m_value;
     };
 
+    /**
+     * @brief Table Entity Data Type.
+     */
+    class TableEntityDataType final {
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:356]: there is no type called ExtensibleEnumeration

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp [hunk 13]
```diff
@@ -378,6 +379,23 @@ namespace Azure { namespace Data { namespace Tables {
...
+     * @brief Initializes a new instance of tableClient.
+     *
+     * @param serviceUrl A url referencing the table that includes the name of the account and the
+     * name of the table.
+     * @param credential The shared key credential used to sign requests.
+     * @param options Optional client options that define the transport pipeline policies for
+     * @param sasBuilder The shared access signature builder used to sign requests.
+     * authentication, retries, etc., that are applied to every request.
+     */
+    explicit TableServicesClient(
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:393]: similar to the storage services we have here two logical entities , one is the service level operations (create/delete table)and one is the table level operations such as add delete row , to documentation also makes this separation

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp (hunk 4 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:418]: done

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp (hunk 5 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:428]: i missed that, sorry  , my bad, done

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp (hunk 6 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:31]: probably not but it is defined as such in the documentation ,other storage service implementations and also in go and .net , thus ..

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    Delete = 8,
+
+    /**
+     * @brief Indicates that all permissions are set.
+     */
+    All = ~0
+  };
+
+  /** @brief Bitwise OR of two values*/
+  inline TablesSasPermissions operator|(TablesSasPermissions lhs, TablesSasPermissions rhs)
```
[the remark this answers is not in the document]
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:68]: what is the macro and i'll use it

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp (hunk 8 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:138]: yes tehy are text, the name of the row / partition, other languages threat them the same, as strings.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp (hunk 7 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp:150]: done

File: sdk/tables/azure-data-tables/samples/tables_getting_started.cpp (hunk 9 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/samples/tables_getting_started.cpp:36]: throw

File: sdk/tables/azure-data-tables/src/tables_clients.cpp (hunk 10 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/src/tables_clients.cpp:138]: done

File: sdk/tables/azure-data-tables/test/ut/sas_test.cpp (hunk 11 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: let m think about it, do we have examples of this pattern ? i suspect w might have something

File: sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp [hunk 15]
```diff
@@ -114,13 +114,13 @@ namespace Azure { namespace Data { namespace Tables { namespace Sas {
     All = ~0,
   };
 
-  inline AccountSasServices operator|(AccountSasServices lhs, AccountSasServices rhs)
+  inline AccountSasServices operator|(AccountSasServices const lhs, AccountSasServices const rhs)
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp:117]: For each type that defines logical operators, could you please define a 7 `constexpr` operators for `^`, `&`, `|`, `~`, `^=`, `&=`, and `|=`, and also have a single unit test per type for each of 7 operations?

Comment (antkmsft) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp:117]: BTW, if you are taking by value, `const` is not needed.

Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp]: done

File: sdk/tables/azure-data-tables/test/ut/sas_test.cpp (hunk 11 above)
Comment (LarryOsterman) [reply sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: The builder pattern is here: https://www.geeksforgeeks.org/builder-pattern-c-design-patterns/ 

Interestingly, this version of the builder pattern and a couple of others does not have the "return *this;" part of the pattern which allows you to chain settings on the builder (other examples I've found do).

The `AccountSasBuilder` is similar to the pattern as described, except it uses mutable fields on the builder rather than methods.

In general, we don't use the Builder pattern in the C++ (we do in other languages like Java and Rust because they're common patterns in those languages). 

IMHO the Storage team mis-used the name "Builder", I would have called them on the use just like I'm calling out the use here.

Humorously, the .Net SDK equivalent uses the "Builder" name, but the Java equivalent does not (Humorously because Java embraces the Builder pattern :))

Because this naming is prevalent in the Storage SDKs as well as .Net (the other P0 languages have different conventions), I can tolerate the name in this case, but its use is questionable elsewhere.

PS: We might want to consider creating an `AccountSasCredential` in `Azure::Core` to express the various SAS credential options for all the services which support SAS credentials (similar to what is done in the other languages)

Comment (gearama) [reply sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: that would be a good step , as this is a common functionality with "flavors" for services, i wonder if we could break it/ templetize it such the the permissions /services enums could be fed in based on the consumer service and generalize the generate method to accomodate the various signature strings and sas parameters

Comment (gearama) [reply sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: filed issue https://github.com/Azure/azure-sdk-for-cpp/issues/5436 to track this and not forget about it

Comment (gearama) [reply sdk/tables/azure-data-tables/test/ut/sas_test.cpp:15]: I agree the name implies a pattern that we end up not respecting. What would be your suggestion , i could spend some time implementing it or rename the objects , again sugestion for the name ?

File: sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  template <class E, class = std::enable_if_t<std::is_enum<E>{}>> E operator&(E lhs, E rhs)
+  {
+    using type = std::underlying_type_t<E>;
+    return static_cast<E>(static_cast<type>(lhs) & static_cast<type>(rhs));
+  }
+
+    /**
+   * @brief Bitwise NOT operator for enum class.
+   */
+  template <class E, class = std::enable_if_t<std::is_enum<E>{}>> E operator!(E val)
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp:30]: unlike operator `~`, `!` is not a bitwise operator, and it should not be here.

**Review by antkmsft — COMMENTED** [review]
What about `|=`, `^=`, and `&=` operators and their tests? Here's how you can write one - https://godbolt.org/z/999EcEY59

File: sdk/core/azure-core/test/ut/string_test.cpp [hunk 17]
```diff
@@ -131,3 +131,52 @@ TEST(String, toUpper)
...
+}
+
+TEST(String, SplitEnd)
+{
+  using Azure::Core::_internal::StringExtensions;
+  std::vector<std::string> result = StringExtensions::Split("1,2,3,", ',');
+  EXPECT_EQ(result[0], "1");
+  EXPECT_EQ(result[1], "2");
+  EXPECT_EQ(result[2], "3");
+  EXPECT_EQ(result.size(), size_t{3});
```
Comment (LarryOsterman) [inline sdk/core/azure-core/test/ut/string_test.cpp:170]: It might make sense for line 170 to be ASSERT_EQ() and move it above the EXPECT_EQ.

That way the test fails immediately on the ASSERT_EQ and we never get to the EXPECT_EQ (which will throw an exception if the ASSERT_EQ is not true).

File: sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+#pragma once
+
+#include <type_traits>
+
+namespace Azure { namespace Data { namespace Tables {
+  /**
+   * @brief Bitwise OR operator for enum class.
+   */
+  template <class E, class = std::enable_if_t<std::is_enum<E>{}>> E operator|(E lhs, E rhs)
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp:12]: `constexpr`?
I'm only marking the first instance, but it can be done to all operators, including the op-and-assign forms.
```suggestion
  template <class E, class = std::enable_if_t<std::is_enum<E>{}>> constexpr E operator|(E lhs, E rhs)
```

File: sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  {
+    using type = std::underlying_type_t<E>;
+    return static_cast<E>(static_cast<type>(lhs) | static_cast<type>(rhs));
+  }
+
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/enum_operators.hpp:17]: please add the `operator~`.
Unlike the `operator!`, the `~` is a bitwise one, and it is useful in expressions like clearing the flag, i.e. `flags &= ~MASK;`

File: sdk/tables/azure-data-tables/test/ut/enum_operators_test.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+// Licensed under the MIT License.
+#include "enum_operators_test.hpp"
+
+using namespace Azure::Data::Tables::Sas;
+using namespace Azure::Data::Tables;
+namespace Azure { namespace Data { namespace Test {
+  TEST(EnumOperator, AndTest)
+  {
+    {
+      auto val = TestEnum::One & TestEnum::Two;
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/test/ut/enum_operators_test.cpp:11]: When you add constexpr before each operator, you can also add at least one per each test for an operator, so that should someone drop `constexpr`, the test would fail to compile.
```suggestion
      constexpr auto val = TestEnum::One & TestEnum::Two;
```

File: sdk/core/azure-core/test/ut/string_test.cpp [hunk 21]
```diff
@@ -131,3 +131,112 @@ TEST(String, toUpper)
...
+  }
+}
+
+TEST(String, SplitSeparator)
+{
+  using Azure::Core::_internal::StringExtensions;
+  {
+    std::vector<std::string> result = StringExtensions::Split(",,,,", ',');
+    ASSERT_EQ(result.size(), size_t{4});
+    EXPECT_EQ(result[0], "");
```
Comment (LarryOsterman) [inline sdk/core/azure-core/test/ut/string_test.cpp:233]: You might want to add a case looking for something that looks like a connection string:

Foo=Bar;Bas=Bletch;Other=SEREWVEWD==

I got burned yesterday because I used Split on (';') and then Split on "=". Which did very bad things on the "Other" field.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp [hunk 22]
```diff
@@ -374,31 +375,70 @@ namespace Azure { namespace Data { namespace Tables {
...
+     * name of the table.
+     * @param credential The shared key credential used to sign requests.
+     * @param options Optional client options that define the transport pipeline policies for
+     * @param sasBuilder The shared access signature builder used to sign requests.
+     * authentication, retries, etc., that are applied to every request.
+     */
+    explicit TableServiceClient(
+        const std::string& serviceUrl,
+        std::shared_ptr<Azure::Data::Tables::Credentials::SharedKeyCredential> credential,
+        Azure::Data::Tables::Sas::AccountSasBuilder& sasBuilder,
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:396]: The sasBuilder is a mutable reference? That's surprising.

Can it be a const reference? And what happens if the sasBuilder's lifetime is different from that of the TableServiceClient? C++ won't catch that kind of lifetime mismatch.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp [hunk 23]
```diff
@@ -202,24 +178,12 @@ namespace Azure { namespace Data { namespace Tables { namespace Sas {
...
-    return static_cast<AccountSasPermissions>(static_cast<type>(lhs) & static_cast<type>(rhs));
-  }
-
   /**
    * @brief AccountSasBuilder is used to generate an account level Shared Access Signature
    * (SAS) for Azure Storage services.
    */
-  struct AccountSasBuilder final
-  {
+  class AccountSasBuilder final {
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp:185]: It is strange that clang-format doesn't complain about this curly brace being on the same line
```suggestion
  class AccountSasBuilder final
  {
```

File: sdk/tables/azure-data-tables/src/tables_clients.cpp [hunk 24]
```diff
@@ -407,6 +418,20 @@ TableClient::TableClient(
       std::move(perOperationPolicies2));
 }
 
+TableClient::TableClient(
+    const std::string& serviceUrl,
+    std::shared_ptr<Azure::Data::Tables::Credentials::SharedKeyCredential> credential,
+    Azure::Data::Tables::Sas::TablesSasBuilder sasBuilder,
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/src/tables_clients.cpp:424]: Why does this take the sasBuilder by value but the other constructor takes it by reference?

It's also feels weird that you take an SasBuilder *and* a SharedKeyCredential.  Looking at the storage clients, they don't seem to follow the same pattern.

Why is Tables different?

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp [hunk 25]
```diff
@@ -347,6 +323,43 @@ namespace Azure { namespace Data { namespace Tables {
...
+       * @brief Construct a new TableEntityDataType object
+       *
+       * @param tableEntityDataType entity data type string.
+       */
+      explicit TableEntityDataType(std::string tableEntityDataType)
+          : ExtendableEnumeration(std::move(tableEntityDataType))
+      {
+      }
+      /** Constant value of type TableEntityDataType:EdmBinary */
+      AZ_DATA_TABLES_DLLEXPORT const static TableEntityDataType EdmBinary;
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:346]: Should we just name them "Binary", "Boolean", "DateTime", i.e. without the "Edm" prefix?

File: sdk/tables/azure-data-tables/test/ut/enum_operators_test.hpp [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+#include "test/ut/test_base.hpp"
+
+namespace Azure { namespace Data { namespace Test {
+  enum class TestEnum
+  {
+    Zero = 0,
+    One = 1,
+    Two = 2,
+    Three = 3,
+    All = ~0
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/test/ut/enum_operators_test.hpp:17]: You might want to consider making the value of `Three` be something other than 3, since you're testing logical operators.

File: sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp [hunk 27]
```diff
@@ -374,31 +375,70 @@ namespace Azure { namespace Data { namespace Tables {
...
+        Core::Context const& context = {});
+
+    /**
+     * @brief Delete the table indicated in the tableName field of the client.
+     *
+     * @param context for canceling long running operations.
+     * @param tableName The name of the table to be deleted.
+     * @return Delete table result.
+     */
+    Response<Models::DeleteResult> DeleteTable(
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/tables_clients.hpp:429]: Should "DeleteResult" be named "DeleteTableResult", or in this context, there is really nothing else you might delete, so "DeleteResult" would always unmistakingly be associated with tables?
I.e. there is no "DeleteTableRow()" method now or in the future, which would also return "DeleteTableRowResult"?

File: sdk/tables/azure-data-tables/src/models.cpp [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+
+  const TableEntityDataType TableEntityDataType::EdmBinary("Edm.Binary");
+  const TableEntityDataType TableEntityDataType::EdmBoolean("Edm.Boolean");
+  const TableEntityDataType TableEntityDataType::EdmDateTime("Edm.DateTime");
+  const TableEntityDataType TableEntityDataType::EdmDouble("Edm.Double");
+  const TableEntityDataType TableEntityDataType::EdmGuid("Edm.Guid");
+  const TableEntityDataType TableEntityDataType::EdmInt32("Edm.Int32");
+  const TableEntityDataType TableEntityDataType::EdmInt64("Edm.Int64");
+  const TableEntityDataType TableEntityDataType::EdmString("Edm.String");
+}}}} // namespace Azure::Data::Tables::Models
```
Comment (antkmsft) [inline sdk/tables/azure-data-tables/src/models.cpp:15]: EOL@EOF
```suggestion
}}}} // namespace Azure::Data::Tables::Models

```

**Review by antkmsft — APPROVED** [review]
I reviewed all changes, LGTM. Thanks for addressing all my previous comments!
I would either get an approval from @LarryOsterman before merging or would generate the next ApiView iteration right after merging, and ask Larry to review.

File: sdk/core/azure-core/test/ut/string_test.cpp (hunk 21 above)
Comment (gearama) [reply sdk/core/azure-core/test/ut/string_test.cpp:233]: one could argue that it did exactly what was expected, even the .net one will remove the == at the end

Comment (gearama) [reply sdk/core/azure-core/test/ut/string_test.cpp:233]: if you don't like it i'll remove it from core and we can continue to spread the split string through the code

File: sdk/tables/azure-data-tables/test/ut/enum_operators_test.hpp (hunk 26 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/test/ut/enum_operators_test.hpp:17]: i already have 0 and all , also those tests are for code coverage more than anything else, i mean what they test is will 1&1 equal 1 when cast to int ?

File: sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp (hunk 25 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/models.hpp:346]: that is the name in the documentation

File: sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp (hunk 23 above)
Comment (gearama) [reply sdk/tables/azure-data-tables/inc/azure/data/tables/account_sas_builder.hpp:185]: keeps formatting it like that , even if i put it like suggested the formatting pulls it back to the previous version

File: sdk/tables/azure-data-tables/inc/azure/data/tables/credentials/azure_sas_credential.hpp [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * @brief Update the signature for the SAS token.
+     */
+    void Update(std::string signature)
+    {
+      std::lock_guard<std::mutex> guard(m_mutex);
+      m_signature = std::move(signature);
+    }
+  };
+}}}} // namespace Azure::Data::Tables::Credentials
```
Comment (LarryOsterman) [inline sdk/tables/azure-data-tables/inc/azure/data/tables/credentials/azure_sas_credential.hpp:45]:
```suggestion
}}}} // namespace Azure::Data::Tables::Credentials

```

**Review by LarryOsterman — APPROVED** [review]
